src/mlkem.rs

```diff
@@ -0,0 +1,371 @@
+
+        mechs.add_mechanism(
+            CKM_ML_KEM_KEY_PAIR_GEN,
+            Self::new_mechanism(CKF_GENERATE_KEY_PAIR),
+        );
+    }
+}
+
+impl Mechanism for MlKemMechanism {
+    fn info(&self) -> &CK_MECHANISM_INFO {
+        &self.info
+    }
+
+    fn encapsulate(
+        &self,
+        _mech: &CK_MECHANISM,
+        key: &Object,
+        key_factory: &Box<dyn ObjectFactory>,
+        template: &[CK_ATTRIBUTE],
+        ciphertext: &mut [u8],
+    ) -> Result<(Object, usize)> {
+        if self.info.flags & CKF_ENCAPSULATE != CKF_ENCAPSULATE {
+            return Err(CKR_MECHANISM_INVALID)?;
+        }
+        match key.check_key_ops(CKO_PUBLIC_KEY, CKK_ML_KEM, CKA_ENCAPSULATE) {
+            Ok(_) => (),
+            Err(e) => return Err(e),
+        }
+        let (keydata, ctlen) = mlkem::encapsulate(key, ciphertext)?;
+        Ok((key_factory.import_from_wrapped(keydata, template)?, ctlen))
+    }
+
+    fn decapsulate(
+        &self,
+        _mech: &CK_MECHANISM,
+        key: &Object,
+        key_factory: &Box<dyn ObjectFactory>,
+        template: &[CK_ATTRIBUTE],
+        ciphertext: &[u8],
+    ) -> Result<Object> {
+        if self.info.flags & CKF_DECAPSULATE != CKF_DECAPSULATE {
+            return Err(CKR_MECHANISM_INVALID)?;
+        }
+        match key.check_key_ops(CKO_PRIVATE_KEY, CKK_ML_KEM, CKA_DECAPSULATE) {
+            Ok(_) => (),
+            Err(e) => return Err(e),
+        }
+        let keydata = mlkem::decapsulate(key, ciphertext)?;
+        key_factory.import_from_wrapped(keydata, template)
+    }
+
+    fn generate_keypair(
+        &self,
+        mech: &CK_MECHANISM,
+        pubkey_template: &[CK_ATTRIBUTE],
+        prikey_template: &[CK_ATTRIBUTE],
+    ) -> Result<(Object, Object)> {
+        let mut pubkey =
+            PUBLIC_KEY_FACTORY.default_object_generate(pubkey_template)?;
+        if !pubkey.check_or_set_attr(Attribute::from_ulong(
+            CKA_CLASS,
+            CKO_PUBLIC_KEY,
+        ))? {
+            return Err(CKR_TEMPLATE_INCONSISTENT)?;
+        }
+        if !pubkey.check_or_set_attr(Attribute::from_ulong(
+            CKA_KEY_TYPE,
+            CKK_ML_KEM,
+        ))? {
+            return Err(CKR_TEMPLATE_INCONSISTENT)?;
+        }
+
+        let param_set = match pubkey.get_attr_as_ulong(CKA_PARAMETER_SET) {
+            Ok(p) => match p {
+                CKP_ML_KEM_512 | CKP_ML_KEM_768 | CKP_ML_KEM_1024 => p,
+                _ => return Err(CKR_ATTRIBUTE_VALUE_INVALID)?,
+            },
+            Err(_) => return Err(CKR_TEMPLATE_INCONSISTENT)?,
+        };
+
+        let mut privkey =
+            PRIVATE_KEY_FACTORY.default_object_generate(prikey_template)?;
+        if !privkey.check_or_set_attr(Attribute::from_ulong(
+            CKA_CLASS,
+            CKO_PRIVATE_KEY,
+        ))? {
+            return Err(CKR_TEMPLATE_INCONSISTENT)?;
+        }
+        if !privkey
+            .check_or_set_attr(Attribute::from_ulong(CKA_KEY_TYPE, CKK_RSA))?
```

Comment:
```suggestion
            .check_or_set_attr(Attribute::from_ulong(CKA_KEY_TYPE, CKK_ML_KEM))?
```
I think we need test coverage :)